src/counterexamples/linear_order_with_pos_mul_pos_eq_zero.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2021 Johan Commelin (inspired by Kevin Buzzard, copied by Damiano Testa).
+All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin (inspired by Kevin Buzzard, copied by Damiano Testa)
+-/
+import algebra.ordered_monoid
+
+/-!
+An example of a `linear_ordered_comm_monoid_with_zero` in which the product of two positive
+elements vanishes.
+
+This is the monoid with 3 elements `0, ε, 1` where `ε ^ 2 = 0` and everything else is forced.
+The order is `0 < ε < 1`.  Since `ε ^ 2 = 0`, the product of strictly positive elements can vanish.
+
+Relevant Zulip chat:
+https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/mul_pos
+-/
+
+/--  The three element monoid. -/
+@[derive [decidable_eq]]
+inductive foo
+| zero
+| eps
+| one
+
+
+namespace foo
+
+instance inhabited : inhabited foo := ⟨zero⟩
+
+instance : has_zero foo := ⟨zero⟩
+instance : has_one foo := ⟨one⟩
+notation `ε` := eps
+
+/-- The order on `foo` is the one induced by the natural order on the image of `aux1`. -/
+def aux1 : foo → ℕ
+| 0 := 0
+| ε := 1
+| 1 := 2
+
+/-- A tactic to prove facts by cases. -/
+meta def boom : tactic unit :=
+`[repeat {rintro ⟨⟩}; dec_trivial]
+
+lemma aux1_inj : function.injective aux1 :=
+by boom
+
+instance : linear_order foo :=
+linear_order.lift aux1 $ aux1_inj
+
+/-- Multiplication on `foo`: the only external input is that `ε ^ 2 = 0`. -/
+def mul : foo → foo → foo
+| 1 x := x
+| x 1 := x
+| _ _ := 0
+
+instance : comm_monoid foo :=
+{ mul := mul,
+  one := 1,
+  one_mul := by boom,
+  mul_one := by boom,
+  mul_comm := by boom,
+  mul_assoc := by boom }
+
+instance : linear_ordered_comm_monoid_with_zero foo :=
+{ zero := 0,
+  zero_mul := by boom,
+  mul_zero := by boom,
+  mul_le_mul_left := by { rintro ⟨⟩ ⟨⟩ h ⟨⟩; revert h; dec_trivial },
+  lt_of_mul_lt_mul_left := by { rintro ⟨⟩ ⟨⟩ ⟨⟩; dec_trivial },
+  zero_le_one := dec_trivial,
+  .. foo.linear_order,
+  .. foo.comm_monoid }
+
+example : 0 < ε ∧ ε * ε = 0 := by boom
```

Comment:
My suggestion above is adding something like:
```suggestion
lemma mul_pos'''' : ¬ ∀ {M : Type*} [linear_ordered_comm_monoid_with_zero M], by exactI ∀
  (a b : M) (ha : 0 < a) (hb : 0 < b), 
  0 < a * b :=
begin
  intro h,
  specialize h ε ε (by boom) (by boom),
  sorry, -- should be easy from here
end
```

Reply:
Eric, I would like to do what you suggest, but I cannot get Lean to process what you sent.  I get the error
```lean
type mismatch at application
  h ε
term
  ε
has type
  foo : Type
but is expected to have type
  ?m_1 : Type u_1
state:
h : ∀ {M : Type u_1} [_inst_1 : linear_ordered_comm_monoid_with_zero M] (a b : M), 0 < a → 0 < b → 0 < a * b
⊢ false
```

and I cannot instantiate the Type M with foo.  I tried for a bit, but nothing seems to work...

Reply:
Oh, that's an interesting problem. My intent was to take all the quantifiers and push them under a `¬`; but I missed one! The universe quantifier can't be moved, and the statement I made above is "this is not true in any universe", whereas probably all you care about proving is "there is a universe in which this is not true".

To prove my statement you need `ulift.up ε` instead of `ε`. But you can probably just replace `Type*` with `Type` in the statement, since the ulift stuff is just noise.

Reply:
Ok, thanks for the explanation!  I had a vague feeling that the `*` in Type was causing problems, but I had no idea how to get around it!

I will try with `ulift.up ε`, and otherwise, I will specialize the Type!

Reply:
I was not able to use `ulift.up ε`, but I removed the `*` from `Type*` and it worked as you suggested!